pkg/utils/kubernetes/networking/services.go

```diff
@@ -54,57 +55,66 @@ func WaitForServiceLoadBalancerAddress(ctx context.Context, c *kubernetes.Client
 			Namespace: namespace,
 			Name:      lbServiceName,
 			Labels: map[string]string{
-				"corresponding-service": name,
+				correspondingSvcNameLabel: name,
 			},
 		},
 		Spec: corev1.ServiceSpec{
 			Type: corev1.ServiceTypeLoadBalancer,
-			// copy the selector and ports of the service to check.
+			// Copy the selector and ports of the service to check.
 			Selector: service.Spec.Selector,
 			Ports:    service.Spec.Ports,
 		},
 	}
 
-	// empty selector, we should create the endpoints separately.
-	// if the target service does not have a selector, it usually means that
+	// Empty selector, we should create the endpoints separately.
+	// If the target service does not have a selector, it usually means that
 	// the endpoints of the target server is manually created, but not chosen from pods by labels in selector.
 	// so we need to manually create the same endpoints as the target service has here.
-	if service.Spec.Selector == nil || len(service.Spec.Selector) == 0 {
-		endpoints, err := c.CoreV1().Endpoints(namespace).Get(ctx, name, metav1.GetOptions{})
+	if len(service.Spec.Selector) == 0 {
+		endpointSlices, err := c.DiscoveryV1().EndpointSlices(namespace).List(
+			ctx, metav1.ListOptions{LabelSelector: discoveryv1.LabelServiceName + "=" + name},
+		)
 		if err != nil {
 			return err
 		}
-		// copy the endpoints for the lb service, and set metadata.
-		tempEndpoints := endpoints.DeepCopy()
-		tempEndpoints.ObjectMeta = metav1.ObjectMeta{
-			Namespace: namespace,
-			Name:      lbServiceName,
-			Labels: map[string]string{
-				"corresponding-service": name,
-			},
-		}
-		_, err = c.CoreV1().Endpoints(namespace).Create(ctx, tempEndpoints, metav1.CreateOptions{})
-		if err != nil {
-			return err
+
+		// Recreate EndpointSlices for the lb service with proper metadata.
+		tempEndpointSlices := endpointSlices.DeepCopy().Items
+		for i := range tempEndpointSlices {
+			tempEndpointSlices[i].ObjectMeta = metav1.ObjectMeta{
+				Namespace: namespace,
+				Name:      fmt.Sprintf("%s-%d", lbServiceName, i),
+				Labels: map[string]string{
+					discoveryv1.LabelServiceName: lbServiceName, // Maps EndpointSlice to Service.
+					correspondingSvcNameLabel:    name,
+				},
+			}
+			_, err = c.DiscoveryV1().EndpointSlices(namespace).Create(ctx, &tempEndpointSlices[i], metav1.CreateOptions{})
+			if err != nil {
+				return err
+			}
 		}
 
 		defer func() {
-			err := c.CoreV1().Endpoints(namespace).Delete(ctx, lbServiceName, metav1.DeleteOptions{})
-			if err != nil && !errors.IsNotFound(err) {
-				fmt.Printf("failed to delete endpoints %s/%s after testing, error %v\n",
-					namespace, lbServiceName, err)
+			for _, eps := range tempEndpointSlices {
+				err := c.DiscoveryV1().EndpointSlices(namespace).Delete(ctx, eps.Name, metav1.DeleteOptions{})
+				if err != nil && !errors.IsNotFound(err) {
+					fmt.Printf("failed to delete endpoints %s/%s after testing, error %v\n",
+						namespace, eps.Name, err,
+					)
+				}
```

Comment:
Shouldn't that whole block land just after the `Create()` above? This way if `i`th `EndpointSlice` where `i < n` and `n` is the length of the `tempEndpointSlices` returns an error upon creation you get the other ones - that were already created - removed.

Reply:
Very, good catch! Thanks, I proposed a solution that ensures every successfully created temporary EndpointSlice will be clean up on return. For each one deffered clean-up function is created, in Go [stack is huge](https://dave.cheney.net/2013/06/02/why-is-a-goroutines-stack-infinite) and I think performance penalty for having multiple `defer` will be negligible in this case